### DISCUSSION TITLE
Can methods on `&self` be called from within other async methods?

### DISCUSSION THREAD

**Discussion by BatmanAoD** [Xuanwo/backon#117] [Q&A] [answered]
I want to implement retry "transparently," within a method that takes `&self`. This is like the ["with self" example](https://github.com/Xuanwo/backon/blob/main/src/docs/examples/with_self.md), except that instead of a `main` method, it uses a member function.

Should something like this work?

```
    async fn invoke(&mut self) -> Result<O, E> {
        use backon::Retryable;
        ( || async {self.invoke_once().await} )
            .retry(backoff)
            .when(self.retry.should_retry)
            .await
    }
```

I'm getting rather intractable trait-bound errors (similar to those in #50).

**Comment by Xuanwo** [comment, marked as answer]
I added a new example here: https://github.com/Xuanwo/backon/pull/118, hoping it will be useful.

You can also paste the entire code and error message so I can offer more assistance.

**Reply by BatmanAoD** [reply]
I apologize; looks like the code I wrote works as-is, and I had confused myself by trying to write a method returning `(&Self, Result<...>)` to accommodate the `Ctx` pattern for `&mut Self`.
